cpp/include/cudf/column/column_device_view.cuh

```diff
@@ -0,0 +1,482 @@
+                                     ///< Based on element type, children
+                                     ///< may contain additional data
+  size_type _num_children{};         ///< The number of child columns
+
+  /**---------------------------------------------------------------------------*
+   * @brief Construct's a `column_device_view` from a `column_view` populating
+   * all but the children.
+   *
+   * @note This constructor is for internal use only. To create a
+   *`column_device_view` from a `column_view`, the
+   *`column_device_view::create()` function should be used.
+   *---------------------------------------------------------------------------**/
+  column_device_view(column_view source);
+
+  /**---------------------------------------------------------------------------*
+   * @brief Destroy the `device_column_view` object.
+   *
+   * @note Does not free the column data, simply free's the device memory
+   * allocated to hold the child views.
+   *---------------------------------------------------------------------------**/
+  void destroy();
+};
+
+/**---------------------------------------------------------------------------*
+ * @brief A mutable, non-owning view of device data as a column of elements
+ * that is trivially copyable and usable CUDA device code.
+ *---------------------------------------------------------------------------**/
+class alignas(16) mutable_column_device_view
+    : public detail::column_device_view_base {
+ public:
+  mutable_column_device_view() = delete;
+  ~mutable_column_device_view() = default;
+  mutable_column_device_view(mutable_column_device_view const&) = default;
+  mutable_column_device_view(mutable_column_device_view&&) = default;
+  mutable_column_device_view& operator=(mutable_column_device_view const&) =
+      default;
+  mutable_column_device_view& operator=(mutable_column_device_view&&) = default;
+
+  /**---------------------------------------------------------------------------*
+   * @brief Factory to construct a column view that is usable in device memory.
+   *
+   * Allocates and copies views of `soure_view`'s children to device memory to
+   * make them accessible in device code.
+   *
+   * If `source_view.num_children() == 0`, then no device memory is allocated.
+   *
+   * Returns a `std::unique_ptr<mutable_column_device_view>` with a custom
+   *deleter to free the device memory allocated for the children.
+   *
+   * A `mutable_column_device_view` should be passed by value into GPU kernels.
+   *
+   * @param source_view The `column_view` to make usable in device code
+   * @param stream optional, stream on which the memory for children will be
+   * allocated
+   * @return A `unique_ptr` to a `mutable_column_device_view` that makes the
+   *data from `source_view` available in device memory.
+   *---------------------------------------------------------------------------**/
+  static auto create(mutable_column_view source_view, cudaStream_t stream = 0);
```

Comment:
Where's the implementation?

Reply:
Implementation is here: https://github.com/rapidsai/cudf/blob/5c1dbd96cf82d07d7ef1e56a467a2920ef695516/cpp/src/column/column_device_view.cu#L39